docs/lbann.rst

```diff
@@ -0,0 +1,72 @@
+********************
+LBANN Semantics
+********************
+
+Trainers (i.e. execution environment)
+******************************************
+
+A trainer is a collection of compute resources and defines a explicit
+communication domain.  It provides the execution for both the training
+and inference of a trained model.  Once constructed a trainer owns an
+LBANN comm object that defines both intra- and inter-trainer
+communication domains.  Additionally, a trainer will contain an I/O
+thread pool that is used to fetch and pre-process data that will be
```

Comment:
While it makes sense for a trainer to own its data readers, it does not make sense for it to own an I/O thread pool. I/O is a shared resource across every trainer resident on a compute node.

Reply:
In that case, we might want to make the thread pool a singleton object.

Reply:
Thread pools are not singletons. That would be a bad move. There can be many thread pools per rank/node/whatever. I/O is the shared resource. IMO, the I/O thread pool is an implementation detail of actually doing I/O and it shouldn't appear in the frontend at all.

Reply:
Good point.  For the moment I don't see a clean way to share a thread pool across multiple trainers.  I will create an issue for later. #1213